File: package/ios/MarkerViewRenderRegistry.swift
import MapKit

/// Splits one native clustering result between SDK objects and live Fabric hosts.
final class MarkerViewRenderRegistry {
  var customClusters = false
  var onChange: ((NativeMarkerViewRenderState) -> Void)? {
    didSet { publish() }
  }
  private var entries: [String: MarkerRenderEntry] = [:]

  var versions: [String: Int] { entries.mapValues { $0.version } }

  func reset() {
    guard !entries.isEmpty else { return }
    entries.removeAll()
    publish()
  }

  func consume(_ diff: MarkerRenderDiff) -> MarkerRenderDiff {
    var removed = diff.removedKeys
    var added: [MarkerRenderEntry] = []
    var retained: [MarkerRenderEntry] = []
    var changed = false
    for key in diff.removedKeys {
      if entries.removeValue(forKey: key) != nil { changed = true }
    }
    func consumeEntry(_ entry: MarkerRenderEntry, isNew: Bool) {
      let live: Bool
      switch entry.element {
      case let .single(descriptor): live = descriptor.customViewId != nil
      case .cluster: live = customClusters
      }
      if live {
        entries[entry.key] = entry
        removed.insert(entry.key)
        changed = true
      } else if entries.removeValue(forKey: entry.key) != nil {
        added.append(entry)
        changed = true
      } else if isNew {
        added.append(entry)
      } else {
        retained.append(entry)
      }
    }
    for entry in diff.added { consumeEntry(entry, isNew: true) }
    for entry in diff.retained { consumeEntry(entry, isNew: false) }
    if changed { publish() }
    return MarkerRenderDiff(removedKeys: removed, added: added, retained: retained)
  }

  private func publish() {
    guard let onChange else { return }
    var ids: [String] = []
    var clusters: [NativeMarkerViewCluster] = []
    for key in entries.keys.sorted() {
      guard let entry = entries[key] else { continue }
      switch entry.element {
      case let .single(descriptor):
        if let id = descriptor.customViewId { ids.append(id) }
      case let .cluster(_, coordinate, _, memberIds, region):
        let longitude = ((coordinate.longitude + 180).truncatingRemainder(dividingBy: 360) + 360)
          .truncatingRemainder(dividingBy: 360) - 180
        clusters.append(NativeMarkerViewCluster(
          id: key,
          coordinate: Coordinate(latitude: coordinate.latitude, longitude: longitude),
          markerIds: memberIds.sorted(),
          region: region.toRegion()
        ))
      }
    }
    onChange(NativeMarkerViewRenderState(markerViewIds: ids, clusters: clusters))
  }
}

File: package/android/src/main/java/com/margelo/nitro/nitromaps/MarkerViewRenderRegistry.kt
package com.margelo.nitro.nitromaps

/** Splits the native cluster display set between SDK objects and Fabric hosts. */
internal class MarkerViewRenderRegistry {
  var customClusters = false
  var onChange: ((NativeMarkerViewRenderState) -> Unit)? = null
    set(value) {
      field = value
      publish()
    }
  private val entries = HashMap<String, ClusterElement>()
  val versions: Map<String, Long> get() = entries.mapValues { it.value.renderVersion }

  fun reset() {
    if (entries.isEmpty()) return
    entries.clear()
    publish()
  }

  fun consume(diff: MarkerRenderDiff): MarkerRenderDiff {
    val removed = diff.removedKeys.toMutableSet()
    val added = ArrayList<ClusterElement>()
    val retained = ArrayList<ClusterElement>()
    var changed = false
    for (key in diff.removedKeys) {
      if (entries.remove(key) != null) changed = true
    }
    fun consumeEntry(element: ClusterElement, isNew: Boolean) {
      val live = when (element) {
        is ClusterElement.Single -> element.descriptor.customViewId != null
        is ClusterElement.Cluster -> customClusters
      }
      if (live) {
        entries[element.diffKey] = element
        removed.add(element.diffKey)
        changed = true
      } else if (entries.remove(element.diffKey) != null) {
        added.add(element)
        changed = true
      } else if (isNew) {
        added.add(element)
      } else {
        retained.add(element)
      }
    }
    diff.added.forEach { consumeEntry(it, true) }
    diff.retained.forEach { consumeEntry(it, false) }
    if (changed) publish()
    return MarkerRenderDiff(removed, added, retained)
  }

  private fun publish() {
    val callback = onChange ?: return
    val ids = ArrayList<String>()
    val clusters = ArrayList<NativeMarkerViewCluster>()
    for (key in entries.keys.sorted()) {
      when (val element = entries[key]) {
        is ClusterElement.Single -> element.descriptor.customViewId?.let { ids.add(it) }
        is ClusterElement.Cluster -> {
          val sw = element.bounds.southwest
          val ne = element.bounds.northeast
          val longitudeSpan = (ne.longitude - sw.longitude + 360.0) % 360.0
          val longitude = ((sw.longitude + longitudeSpan / 2 + 540.0) % 360.0) - 180.0
          clusters.add(NativeMarkerViewCluster(
            key,
            Coordinate(element.position.latitude, element.position.longitude),
            element.memberIds.sorted().toTypedArray(),
            Region((sw.latitude + ne.latitude) / 2, longitude,
              (ne.latitude - sw.latitude).coerceAtLeast(0.0001), longitudeSpan.coerceAtLeast(0.0001)),
          ))
        }
        null -> Unit
      }
    }
    callback(NativeMarkerViewRenderState(ids.toTypedArray(), clusters.toTypedArray()))
  }
}
